File: packages/react-native-autoplay/ios/hybrid/HybridMapTemplate.swift
//
//  HybridMapTemplate.swift
//  Pods
//
//  Created by Manuel Auer on 15.10.25.
//

class HybridMapTemplate: HybridHybridMapTemplateSpec {
    func createMapTemplate(config: MapTemplateConfig) throws {
        let template = MapTemplate(config: config)
        try RootModule.withScene { scene in
            scene.templateStore.addTemplate(
                template: template,
                templateId: config.id
            )
        }
    }

    func setTemplateMapButtons(templateId: String, buttons: [NitroMapButton]?)
        throws
    {
        try RootModule.withTemplate(templateId: templateId) {
            template in
            guard let template = template as? MapTemplate else {
                return
            }

            template.config.mapButtons = buttons
            template.invalidate()
        }
    }

    func showNavigationAlert(templateId: String, alert: NitroNavigationAlert)
        throws -> NavigationAlertCallbacks
    {
        var mapTemplate: MapTemplate?

        try RootModule.withMapTemplate(templateId: templateId) { template in
            template.showAlert(alertConfig: alert)
            mapTemplate = template
        }

        return NavigationAlertCallbacks(
            dismiss: {
                mapTemplate?.dismissNavigationAlert(alertId: alert.id)
            },
            update: { title, subtitle in
                mapTemplate?.updateNavigationAlert(
                    alertId: alert.id,
                    title: title,
                    subtitle: subtitle
                )
            }
        )
    }

    func showTripSelector(
        templateId: String,
        trips: [TripsConfig],
        selectedTripId: String?,
        textConfig: TripPreviewTextConfiguration,
        onTripSelected: @escaping (_ tripId: String, _ routeId: String) -> Void,
        onTripStarted: @escaping (_ tripId: String, _ routeId: String) -> Void,
        onBackPressed: @escaping () -> Void,
        mapButtons: [NitroMapButton]
    ) throws -> TripSelectorCallback {
        var callback: TripSelectorCallback?

        try RootModule.withMapTemplate(templateId: templateId) { template in
            callback = try template.showTripSelector(
                trips: trips,
                selectedTripId: selectedTripId,
                textConfig: textConfig,
                onTripSelected: onTripSelected,
                onTripStarted: onTripStarted,
                onBackPressed: onBackPressed,
                mapButtons: mapButtons
            )
        }

        guard let callback = callback else {
            throw AutoPlayError.templateNotFound(templateId)
        }

        return callback
    }

    func hideTripSelector(templateId: String) throws {
        try RootModule.withMapTemplate(templateId: templateId) { template in
            template.hideTripSelector()
        }
    }

    func updateVisibleTravelEstimate(
        templateId: String,
        visibleTravelEstimate: VisibleTravelEstimate
    ) throws {
        try RootModule.withMapTemplate(templateId: templateId) { template in
            template.updateVisibleTravelEstimate(
                visibleTravelEstimate: visibleTravelEstimate
            )
        }
    }

    func updateTravelEstimates(templateId: String, steps: [TripPoint]) throws {
        try RootModule.withMapTemplate(templateId: templateId) { template in
            template.updateTravelEstimates(steps: steps)
        }
    }

    func updateManeuvers(templateId: String, maneuvers: NitroManeuver) throws {
        try RootModule.withMapTemplate(templateId: templateId) {
            template in
            switch maneuvers {
            case .first(let routingManeuvers):
                {
                    template.updateManeuvers(maneuvers: routingManeuvers)
                }()
            case .second(let messageManeuver):
                {
                    template.updateManeuvers(messageManeuver: messageManeuver)
                }()
            }

        }
    }

    func startNavigation(templateId: String, trip: TripConfig) throws {
        try RootModule.withMapTemplate(templateId: templateId) { template in
            let trip = Parser.parseTrip(tripConfig: trip)
            template.startNavigation(trip: trip)
        }
    }

    func stopNavigation(templateId: String) throws {
        try RootModule.withMapTemplate(templateId: templateId) { template in
            template.stopNavigation()
        }
    }

}

File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/swe/iternio/reactnativeautoplay/HybridMapTemplate.kt
package com.margelo.nitro.swe.iternio.reactnativeautoplay

import androidx.car.app.AppManager
import com.facebook.react.bridge.UiThreadUtil
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.AndroidAutoTemplate
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.MapTemplate
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.RoutePreviewTemplate
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.TripPreviewTemplate

class HybridMapTemplate : HybridMapTemplateSpec() {
    var navigationAlert: NitroNavigationAlert? = null

    override fun createMapTemplate(config: MapTemplateConfig) {
        val context = AndroidAutoSession.getCarContext(config.id) ?: throw IllegalArgumentException(
            "createMapTemplate failed, carContext found"
        )

        val template = MapTemplate(context, config, initNavigationManager = true)
        AndroidAutoTemplate.setTemplate(config.id, template)
    }

    override fun showNavigationAlert(
        templateId: String, alert: NitroNavigationAlert
    ) {
        val alertConfig = alert.copy(onDidDismiss = { reason ->
            navigationAlert = null
            alert.onDidDismiss?.let { it(reason) }
        })
        val template = AndroidAutoTemplate.getTemplate<MapTemplate>(templateId)
        template.showAlert(alertConfig)
        navigationAlert = alertConfig
    }

    override fun updateNavigationAlert(
        templateId: String, navigationAlertId: Double, title: AutoText, subtitle: AutoText?
    ) {
        navigationAlert?.let {
            if (it.id != navigationAlertId) {
                return
            }
            val navigationAlert = it.copy(title = title, subtitle = subtitle)
            val template = AndroidAutoTemplate.getTemplate<MapTemplate>(templateId)
            template.showAlert(navigationAlert)
        }
    }

    override fun dismissNavigationAlert(
        templateId: String, navigationAlertId: Double
    ) {
        val carContext = AndroidAutoSession.getCarContext(AndroidAutoSession.ROOT_SESSION)
            ?: throw IllegalArgumentException(
                "dismissNavigationAlert failed, carContext found"
            )
        carContext.getCarService(AppManager::class.java).dismissAlert(navigationAlertId.toInt())
    }

    override fun showTripSelector(
        templateId: String,
        trips: Array<TripsConfig>,
        selectedTripId: String?,
        textConfig: TripPreviewTextConfiguration,
        onTripSelected: (tripId: String, routeId: String) -> Unit,
        onTripStarted: (tripId: String, routeId: String) -> Unit,
        onBackPressed: () -> Unit,
        mapButtons: Array<NitroMapButton>
    ): TripSelectorCallback {
        val context = AndroidAutoSession.getRootContext()
            ?: throw IllegalArgumentException("showTripSelector failed, carContext not found")
        val screenManager = AndroidAutoScreen.getScreenManager()
            ?: throw IllegalArgumentException("showTripSelector failed, screenManager not found")

        val screen = TripPreviewTemplate(
            context,
            trips,
            selectedTripId,
            textConfig,
            onTripSelected,
            onTripStarted,
            onBackPressed,
            mapButtons
        )

        UiThreadUtil.runOnUiThread {
            val topMessageTemplate = AndroidAutoTemplate.getTopMessageTemplate()

            screenManager.push(screen)

            topMessageTemplate?.let {
                // make sure any visible message template is still on top of the trip selector
                screenManager.push(topMessageTemplate)
            }
        }

        return TripSelectorCallback { id: String ->
            UiThreadUtil.runOnUiThread {
                if (screenManager.top.marker == RoutePreviewTemplate.TAG) {
                    screenManager.popTo(TripPreviewTemplate.TAG)
                }
            }

            val selectedTripIndex = trips.indexOfFirst { trip -> trip.id == id }
            screen.selectedTripIndex = selectedTripIndex
            screen.invalidate()
        }
    }

    override fun hideTripSelector(templateId: String) {
        val screenManager = AndroidAutoScreen.getScreenManager()
            ?: throw IllegalArgumentException("hideTripSelector failed, screenManager not found")
        val screens = screenManager.screenStack.filter {
            it.marker == TripPreviewTemplate.TAG || it.marker == RoutePreviewTemplate.TAG
        }

        UiThreadUtil.runOnUiThread {
            screens.forEach {
                it.finish()
            }
        }
    }

    override fun setTemplateMapButtons(
        templateId: String, buttons: Array<NitroMapButton>?
    ) {
        val template = AndroidAutoTemplate.getTemplate<MapTemplate>(templateId)
        template.setMapActions(buttons)
    }

    override fun updateVisibleTravelEstimate(
        templateId: String, visibleTravelEstimate: VisibleTravelEstimate
    ) {
        val template = AndroidAutoTemplate.getTemplate<MapTemplate>(templateId)
        template.updateVisibleTravelEstimate(visibleTravelEstimate)
    }

    override fun updateTravelEstimates(templateId: String, steps: Array<TripPoint>) {
        MapTemplate.updateTravelEstimates(steps)
    }

    override fun updateManeuvers(
        templateId: String, maneuvers: NitroManeuver
    ) {
        MapTemplate.updateManeuvers(maneuvers)
    }

    override fun startNavigation(
        templateId: String, trip: TripConfig
    ) {
        MapTemplate.startNavigation(trip)
    }

    override fun stopNavigation(templateId: String) {
        MapTemplate.stopNavigation()
    }
}